File: src/main/java/de/twometer/neko/gl/Window.java
package de.twometer.neko.gl;

import de.twometer.neko.event.*;
import de.twometer.neko.res.AssetPaths;
import de.twometer.neko.res.ResourceLoader;
import de.twometer.neko.util.Cache;
import de.twometer.neko.util.CrashHandler;
import org.joml.Vector2f;
import org.lwjgl.glfw.*;
import org.lwjgl.opengl.GL;

import java.io.IOException;
import java.util.Objects;

import static org.lwjgl.glfw.Callbacks.glfwFreeCallbacks;
import static org.lwjgl.glfw.GLFW.*;
import static org.lwjgl.opengl.GL11.GL_TRUE;
import static org.lwjgl.system.MemoryUtil.NULL;

public class Window {

    private long handle;

    private final String title;

    private int width;

    private int height;

    private float scale;

    private final Cache<Integer, Long> cursorCache = new Cache<>() {
        @Override
        protected Long create(Integer integer) {
            return glfwCreateStandardCursor(integer);
        }
    };

    public Window(String title, int width, int height) {
        this.title = title;
        this.width = width;
        this.height = height;
    }

    public void setCursorVisible(boolean visible) {
        glfwSetInputMode(handle, GLFW_CURSOR, visible ? GLFW_CURSOR_NORMAL : GLFW_CURSOR_DISABLED);
    }

    public boolean isCursorVisible() {
        return glfwGetInputMode(handle, GLFW_CURSOR) == GLFW_CURSOR_NORMAL;
    }

    public void create() {
        GLFWErrorCallback.createPrint(System.err).set();

        if (!glfwInit())
            throw new IllegalStateException("Failed to initialize GLFW");

        glfwWindowHint(GLFW_CONTEXT_VERSION_MAJOR, 3);
        glfwWindowHint(GLFW_CONTEXT_VERSION_MINOR, 3);
        glfwWindowHint(GLFW_OPENGL_FORWARD_COMPAT, GL_TRUE);
        glfwWindowHint(GLFW_OPENGL_PROFILE, GLFW_OPENGL_CORE_PROFILE);

        handle = glfwCreateWindow(width, height, title, NULL, NULL);
        if (handle == NULL)
            throw new IllegalStateException("Failed to create GLFW window");

        glfwMakeContextCurrent(handle);
        glfwSwapInterval(1);

        float[] scaleBuf = new float[1];
        glfwGetWindowContentScale(handle, scaleBuf, scaleBuf);
        scale = scaleBuf[0];

        if (scale > 1.0f)
            setSize((int) (width * scale), (int) (height * scale));

        GL.createCapabilities();

        glfwSetFramebufferSizeCallback(handle, (window, width, height) -> {
            this.width = width;
            this.height = height;

            Events.post(new SizeChangedEvent(width, height));
        });

        glfwSetMouseButtonCallback(handle, (window, button, action, mods) -> {
            if (action == GLFW_RELEASE)
                Events.post(new MouseClickedEvent(button));
            Events.post(new MouseButtonEvent(button, action, mods));
        });

        glfwSetKeyCallback(handle, (window, key, scancode, action, mods) -> {
            if (action == GLFW_REPEAT || action == GLFW_PRESS)
                Events.post(new KeyPressedEvent(key));
            Events.post(new KeyEvent(key, scancode, action, mods));
        });
        glfwSetCharCallback(handle, (window, codepoint) -> Events.post(new CharTypedEvent((char) codepoint, codepoint)));
        glfwSetCursorPosCallback(handle, (window, xpos, ypos) -> Events.post(new MousePosEvent((int) xpos, (int) ypos)));
        glfwSetScrollCallback(handle, (window, xoffset, yoffset) -> Events.post(new ScrollEvent((int) xoffset, (int) yoffset)));
        glfwSetWindowFocusCallback(handle, (window, focused) -> Events.post(new WindowFocusEvent(focused)));
    }

    public void update() {
        glfwSwapBuffers(handle);
        glfwPollEvents();
    }

    public void destroy() {
        glfwFreeCallbacks(handle);
        glfwDestroyWindow(handle);

        glfwTerminate();
        Objects.requireNonNull(glfwSetErrorCallback(null)).free();
    }

    public float getScale() {
        return scale;
    }

    public boolean shouldClose() {
        return glfwWindowShouldClose(handle);
    }

    public void close() {
        glfwSetWindowShouldClose(handle, true);
    }

    public void setCursorPosition(Vector2f vec) {
        glfwSetCursorPos(handle, vec.x, vec.y);
    }

    public boolean isKeyPressed(int key) {
        return glfwGetKey(handle, key) == GLFW_PRESS;
    }

    public boolean isMouseButtonPressed(int key) {
        return glfwGetMouseButton(handle, key) == GLFW_PRESS;
    }

    public Vector2f getCursorPosition() {
        double[] xPos = new double[1];
        double[] yPos = new double[1];
        glfwGetCursorPos(handle, xPos, yPos);
        return new Vector2f((float) xPos[0], (float) yPos[0]);
    }

    private void setSize(int width, int height) {
        this.width = width;
        this.height = height;
        glfwSetWindowSize(handle, width, height);
    }

    public int getWidth() {
        return width;
    }

    public int getHeight() {
        return height;
    }

    public String getClipboardContent() {
        return glfwGetClipboardString(handle);
    }

    public void setClipboardContent(String str) {
        glfwSetClipboardString(handle, str);
    }

    public void setCursor(int cursor) {
        var cursorObj = cursor == 0 ? 0 : cursorCache.get(cursor);
        glfwSetCursor(handle, cursorObj);
    }

    public void setIcon(String path) {
        try {
            var img = ResourceLoader.loadImage(AssetPaths.TEXTURE_PATH + path);
            var pix = ResourceLoader.loadPixels(img);
            var buf = GLFWImage.malloc(1);
            var glfwImg = GLFWImage.malloc();
            glfwImg.set(img.getWidth(), img.getHeight(), pix);
            buf.put(0, glfwImg);
            glfwSetWindowIcon(handle, buf);
        } catch (IOException e) {
            CrashHandler.fatal(e);
        }
    }

}

File: neko-engine/src/main/kotlin/de/twometer/neko/core/Window.kt
package de.twometer.neko.core

import de.twometer.neko.events.*
import de.twometer.neko.res.AssetManager
import de.twometer.neko.res.AssetType
import de.twometer.neko.res.ImageLoader
import de.twometer.neko.util.Cache
import org.lwjgl.glfw.Callbacks.glfwFreeCallbacks
import org.lwjgl.glfw.GLFW.*
import org.lwjgl.glfw.GLFWErrorCallback
import org.lwjgl.glfw.GLFWImage
import org.lwjgl.opengl.GL
import org.lwjgl.system.MemoryUtil.NULL


class Window(private val config: AppConfig) {

    var handle: Long = 0
        private set

    private var cursorVisible: Boolean = true

    private val cursorCache = object : Cache<Int, Long>() {
        override fun create(key: Int): Long = glfwCreateStandardCursor(key)
    }

    fun create() {
        GLFWErrorCallback.createPrint(System.err).set()

        if (!glfwInit())
            error("GLFW init failed")

        glfwWindowHint(GLFW_CONTEXT_VERSION_MAJOR, config.glMajor)
        glfwWindowHint(GLFW_CONTEXT_VERSION_MINOR, config.glMinor)
        glfwWindowHint(GLFW_OPENGL_FORWARD_COMPAT, GLFW_TRUE)
        glfwWindowHint(GLFW_OPENGL_PROFILE, GLFW_OPENGL_CORE_PROFILE)

        handle = glfwCreateWindow(config.windowWidth, config.windowHeight, config.windowTitle, 0, 0)
        if (handle == NULL)
            error("GLFW window creation failed")

        glfwMakeContextCurrent(handle)
        glfwSwapInterval(1)

        GL.createCapabilities()

        glfwSetFramebufferSizeCallback(handle) { _, width, height -> Events.post(ResizeEvent(width, height)) }
        glfwSetWindowFocusCallback(handle) { _, focused -> Events.post(FocusEvent(focused)) }
        glfwSetCharCallback(handle) { _, codepoint -> Events.post(CharEvent(codepoint.toChar(), codepoint)) }
        glfwSetCursorPosCallback(handle) { _, x, y -> Events.post(MouseMoveEvent(x.toInt(), y.toInt())) }
        glfwSetScrollCallback(handle) { _, x, y -> Events.post(MouseWheelEvent(x.toInt(), y.toInt())) }

        glfwSetMouseButtonCallback(handle) { _, button, action, mods ->
            if (action == GLFW_RELEASE)
                Events.post(MouseClickEvent(button))
            Events.post(MouseButtonEvent(button, action, mods))
        }

        glfwSetKeyCallback(handle) { _, key, scancode, action, mods ->
            if (action == GLFW_REPEAT || action == GLFW_PRESS)
                Events.post(KeyPressEvent(key))
            Events.post(KeyEvent(key, scancode, action, mods))
        }

        if (config.windowIcon != null) {
            val image = ImageLoader.load(AssetManager.resolve(config.windowIcon, AssetType.Textures).absolutePath)
            val glfwArray = GLFWImage.malloc(1)
            val glfwImage = GLFWImage.malloc()
            glfwImage.set(image.width, image.height, image.pixels)
            glfwArray.put(0, glfwImage)
            glfwSetWindowIcon(handle, glfwArray)
        }
    }

    fun update() {
        glfwSwapBuffers(handle)
        glfwPollEvents()
    }

    fun destroy() {
        if (handle == NULL)
            return

        cursorCache.map().forEach { glfwDestroyCursor(it.value) }
        glfwFreeCallbacks(handle)
        glfwDestroyWindow(handle)
        glfwTerminate()
        glfwSetErrorCallback(null)?.free()
        GL.destroy()
    }

    fun setSize(width: Int, height: Int) = glfwSetWindowSize(handle, width, height)

    fun setTitle(title: String) = glfwSetWindowTitle(handle, title)

    fun setCursorVisible(visible: Boolean) {
        if (visible != cursorVisible) {
            glfwSetInputMode(handle, GLFW_CURSOR, if (visible) GLFW_CURSOR_NORMAL else GLFW_CURSOR_DISABLED)
            cursorVisible = visible
        }
    }

    fun setCursorPosition(x: Int, y: Int) = glfwSetCursorPos(handle, x.toDouble(), y.toDouble())

    fun setCursor(cursor: Int) = glfwSetCursor(handle, if (cursor == 0) 0 else cursorCache.get(cursor))

    fun setClipboardContent(str: String) {
        glfwSetClipboardString(handle, str)
    }

    fun getClipboardContent(): String? {
        return glfwGetClipboardString(handle)
    }

    fun getScale(): Float {
        val f = floatArrayOf(0.0f)
        glfwGetWindowContentScale(handle, f, f)
        return f[0]
    }

    fun getSize(): Pair<Int, Int> {
        val w = intArrayOf(0)
        val h = intArrayOf(0)
        glfwGetFramebufferSize(handle, w, h)
        return Pair(w[0], h[0])
    }

    fun isCursorEnabled() = glfwGetInputMode(handle, GLFW_CURSOR) == GLFW_CURSOR_NORMAL

    fun getCursorPosition(): Pair<Double, Double> {
        val x = doubleArrayOf(0.0)
        val y = doubleArrayOf(0.0)
        glfwGetCursorPos(handle, x, y)
        return Pair(x[0], y[0])
    }

    fun isFocused() = glfwGetWindowAttrib(handle, GLFW_FOCUSED) == GLFW_TRUE

    fun isCloseRequested(): Boolean = glfwWindowShouldClose(handle)

    fun isKeyDown(key: Int) =
        glfwGetKey(handle, key) == GLFW_PRESS

    fun isMouseButtonPressed(key: Int) = glfwGetMouseButton(handle, key) == GLFW_PRESS

    fun close() {
        glfwSetWindowShouldClose(handle, true)
    }

}